### PR TITLE
BUG: fixes bug when using sep=None and comment keyword for read_csv

### LINKED ISSUE #31396: TypeError when using 'comment=...' in read_csv from a file
#### Code Sample

Given a data file `data.csv` with a line that is commented out:
```csv
+1.280000e+002,-4.078996e+001
+2.560000e+002,-5.155923e+001
# +3.840000e+002,-7.221378e+001
+5.120000e+002,-7.918677e+001
+6.400000e+002,-7.919656e+001
```
```python
import pandas as pd
pd.read_csv('data.csv', sep=None, index_col=0, header=None, engine="python", comment='#')
```
#### Problem description

It raises a `TypeError` when using the `comment` parameter:

<details>
<summary>TypeError</summary>

```python
TypeError                                 Traceback (most recent call last)
<ipython-input-17-c89b3c3e691f> in <module>
----> 1 pd.read_csv('data.csv', sep=None, comment='#')

~/.local/share/virtualenvs/openqlab/lib/python3.8/site-packages/pandas/io/parsers.py in parser_f(filepath_or_buffer, sep, delimiter, header, names,
 index_col, usecols, squeeze, prefix, mangle_dupe_cols, dtype, engine, converters, true_values, false_values, skipinitialspace, skiprows, skipfoote
r, nrows, na_values, keep_default_na, na_filter, verbose, skip_blank_lines, parse_dates, infer_datetime_format, keep_date_col, date_parser, dayfirs
t, cache_dates, iterator, chunksize, compression, thousands, decimal, lineterminator, quotechar, quoting, doublequote, escapechar, comment, encodin
g, dialect, error_bad_lines, warn_bad_lines, delim_whitespace, low_memory, memory_map, float_precision)
    683         )
    684 
--> 685         return _read(filepath_or_buffer, kwds)
    686 
    687     parser_f.__name__ = name

~/.local/share/virtualenvs/openqlab/lib/python3.8/site-packages/pandas/io/parsers.py in _read(filepath_or_buffer, kwds)
    455 
    456     # Create the parser.
--> 457     parser = TextFileReader(fp_or_buf, **kwds)
    458 
    459     if chunksize or iterator:

~/.local/share/virtualenvs/openqlab/lib/python3.8/site-packages/pandas/io/parsers.py in __init__(self, f, engine, **kwds)
    893             self.options["has_index_names"] = kwds["has_index_names"]
    894 
--> 895         self._make_engine(self.engine)
    896 
    897     def close(self):

~/.local/share/virtualenvs/openqlab/lib/python3.8/site-packages/pandas/io/parsers.py in _make_engine(self, engine)
   1145                     ' "python-fwf")'.format(engine=engine)
   1146                 )
-> 1147             self._engine = klass(self.f, **self.options)
   1148 
   1149     def _failover_to_python(self):

~/.local/share/virtualenvs/openqlab/lib/python3.8/site-packages/pandas/io/parsers.py in __init__(self, f, **kwds)
   2297         # Set self.data to something that can read lines.
   2298         if hasattr(f, "readline"):
-> 2299             self._make_reader(f)
   2300         else:
   2301             self.data = f

~/.local/share/virtualenvs/openqlab/lib/python3.8/site-packages/pandas/io/parsers.py in _make_reader(self, f)
   2427                 self.pos += 1
   2428                 self.line_pos += 1
-> 2429                 sniffed = csv.Sniffer().sniff(line)
   2430                 dia.delimiter = sniffed.delimiter
   2431                 if self.encoding is not None:

/usr/lib64/python3.8/csv.py in sniff(self, sample, delimiters)
    179 
    180         quotechar, doublequote, delimiter, skipinitialspace = \
--> 181                    self._guess_quote_and_delimiter(sample, delimiters)
    182         if not delimiter:
    183             delimiter, skipinitialspace = self._guess_delimiter(sample,

/usr/lib64/python3.8/csv.py in _guess_quote_and_delimiter(self, data, delimiters)
    220                       r'(?:^|\n)(?P<quote>["\']).*?(?P=quote)(?:$|\n)'):                            #  ".*?" (no delim, no space)
    221             regexp = re.compile(restr, re.DOTALL | re.MULTILINE)
--> 222             matches = regexp.findall(data)
    223             if matches:
    224                 break

TypeError: expected string or bytes-like object
```

</details>

Without the comment in the data file and without the parameter `comment='#'` everything works as expected.

**It seems that `sep=None` ist the problem here.**
When using `sep=','` it works. But in our case, the import is part of a general importer that should accept a variety of different files. Thus, we must use `sep=None`. 

#### Expected Output

I would expect the following output:

```python
Out[18]: 
                 1
0                 
128.0    -40.78996
256.0    -51.55923
512.0    -79.18677
640.0    -79.19656

[4 rows x 1 columns]
```

#### Output of ``pd.show_versions()``

<details>
<summary>Details</summary>

INSTALLED VERSIONS
------------------
commit           : None
python           : 3.8.1.final.0
python-bits      : 64
OS               : Linux
OS-release       : 5.4.13-201.fc31.x86_64
machine          : x86_64
processor        : x86_64
byteorder        : little
LC_ALL           : None
LANG             : de_DE.UTF-8
LOCALE           : de_DE.UTF-8

pandas           : 0.25.3
numpy            : 1.18.1
pytz             : 2019.3
dateutil         : 2.8.1
pip              : 19.3.1
setuptools       : 42.0.2
Cython           : None
pytest           : 5.3.2
hypothesis       : None
sphinx           : 2.3.1
blosc            : None
feather          : None
xlsxwriter       : None
lxml.etree       : None
html5lib         : None
pymysql          : None
psycopg2         : None
jinja2           : 2.10.3
IPython          : 7.11.1
pandas_datareader: None
bs4              : None
bottleneck       : None
fastparquet      : None
gcsfs            : None
lxml.etree       : None
matplotlib       : 3.1.2
numexpr          : None
odfpy            : None
openpyxl         : None
pandas_gbq       : None
pyarrow          : None
pytables         : None
s3fs             : None
scipy            : 1.4.1
sqlalchemy       : None
tables           : None
xarray           : None
xlrd             : None
xlwt             : None
xlsxwriter       : None

</details>

### PR DESCRIPTION
**Pull request by s-scherrer**
- [x] closes #31396 
- [x] tests added / passed
- [x] passes `black pandas`
- [x] passes `git diff upstream/master -u -- "*.py" | flake8 --diff`
- [x] whatsnew entry - added

### CODE REVIEW

File: pandas/tests/io/parser/test_python_parser_only.py [hunk 1]
```diff
@@ -66,6 +66,23 @@ def test_sniff_delimiter(python_parser_only, kwargs):
     tm.assert_frame_equal(result, expected)
 
 
+def test_sniff_delimiter_comment(python_parser_only):
```
Comment (jreback) [inline pandas/tests/io/parser/test_python_parser_only.py]: can you move this near other comment tests. & add the issue number as a comment

**Review by jreback — CHANGES_REQUESTED** [review]
can you add a whatsnew note. let's target 1.1 with this.

File: pandas/tests/io/parser/test_python_parser_only.py (hunk 1 above)
Comment (s-scherrer) [reply pandas/tests/io/parser/test_python_parser_only.py]: I moved the test to the end of the file, is this what you meant? Or would you rather have me put it in tests/io/parser/test_comment.py?

Comment (gfyoung) [reply pandas/tests/io/parser/test_python_parser_only.py]: `test_comment.py`

File: doc/source/whatsnew/v1.1.0.rst [hunk 2]
```diff
@@ -180,7 +180,7 @@ I/O
 - Bug in :meth:`read_json` where integer overflow was occuring when json contains big number strings. (:issue:`30320`)
 - `read_csv` will now raise a ``ValueError`` when the arguments `header` and `prefix` both are not `None`. (:issue:`27394`)
 - Bug in :meth:`DataFrame.to_json` was raising ``NotFoundError`` when ``path_or_buf`` was an S3 URI (:issue:`28375`)
--
+- Bug in :meth:`read_csv` was raising `TypeError` when `sep=None` was used in combination with `comment` keyword (:issue:`31396`)
```
Comment (gfyoung) [inline doc/source/whatsnew/v1.1.0.rst]: Please resolve the merge conflict.

File: pandas/tests/io/parser/test_python_parser_only.py (hunk 1 above)
Comment (s-scherrer) [reply pandas/tests/io/parser/test_python_parser_only.py]: Should I move the full test or simply add `dict(sep=None)` as parameter in `test_line_comment` and then skip the c-enginge tests like it's done for the lineterminator test?

**Comment by s-scherrer** [comment]
When adding `dict(sep=None)` as parameter to `test_line_comment` I found that my current fix does not work if the first line is a comment.

Overall I think the flow of comment removal and separating columns is a bit inconsistent. During searching for the bug I found that `PythonParser._check_comments` is sometimes called before separating the line (e.g. when sniffing the separator) and sometimes afterwards. In my opinion it would make more sense  to remove comments first in all cases.

File: pandas/tests/io/parser/test_python_parser_only.py (hunk 1 above)
Comment (gfyoung) [reply pandas/tests/io/parser/test_python_parser_only.py]: Move the full test as is and change the fixture from `python_parser_only` to `all_parsers`.

Comment (s-scherrer) [reply pandas/tests/io/parser/test_python_parser_only.py]: This issue affects only `PythonParser`, `sep=None` does not work with the c engine.

Comment (gfyoung) [reply pandas/tests/io/parser/test_python_parser_only.py]: Oh, I see...

Actually, never mind then.  Per the directive in this file, this test is purely for Python parser-only tests, with the goal of migrating them out of here at later date.

Let's keep the test here in fact.

**Comment by s-scherrer** [comment]
The Travis CI build failed with the following message:
```
UnavailableInvalidChannel: The channel is not accessible or is invalid.

  channel name: c3i_test

  channel url: https://conda.anaconda.org/c3i_test

  error code: 404
```

Is there a way to rerun it?

**Comment by gfyoung** [comment]
@s-scherrer : This is actually an unrelated error that we need to resolve on `master`.

cc @pandas-dev/pandas-core

**Review by WillAyd — APPROVED** [review]
lgtm

**Review by jreback — APPROVED** [review]

**Comment by jreback** [comment]
thanks @s-scherrer

**Pull request merged**